gateway-ha/src/main/java/io/trino/gateway/ha/router/ProcessedRequest.java

```diff
@@ -0,0 +1,533 @@
+import java.util.Optional;
+import java.util.Set;
+import java.util.stream.Collectors;
+
+import static com.google.common.io.BaseEncoding.base64Url;
+import static io.airlift.json.JsonCodec.jsonCodec;
+import static java.lang.Math.toIntExact;
+import static java.nio.charset.StandardCharsets.UTF_8;
+import static java.util.Objects.requireNonNull;
+import static java.util.Objects.requireNonNullElse;
+
+public class ProcessedRequest
+{
+    public static final String UNSET = "__UNSET__";
+
+    private final Logger log = Logger.get(ProcessedRequest.class);
+    private String oauthTokenInfoUrl;
+    private final boolean isGetUserFromOAuth;
+    private final boolean isClientsUseV2Format;
+    private String body = "";
+    private String queryType = "";
+    private Set<QualifiedName> tables = ImmutableSet.of();
+    private final String defaultCatalog;
+    private final String defaultSchema;
+    private Set<String> catalogs = ImmutableSet.of();
+    private Set<String> schemas = ImmutableSet.of();
+    private Set<String> catalogSchemas = ImmutableSet.of();
+    private Optional<String> user = Optional.empty();
+
+    public static final String TRINO_USER_HEADER_NAME = "X-Trino-User";
+    public static final String TRINO_CATALOG_HEADER_NAME = "X-Trino-Catalog";
+    public static final String TRINO_SCHEMA_HEADER_NAME = "X-Trino-Schema";
+    public static final String TRINO_PREPARED_STATEMENT_HEADER_NAME = "X-Trino-Prepared-Statement";
+
+    public static final String TRINO_UI_TOKEN_NAME = "Trino-UI-Token";
+    public static final String TRINO_SECURE_UI_TOKEN_NAME = "__Secure-Trino-ID-Token";
+
+    public ProcessedRequest(HttpServletRequest request, ProcessedRequestConfig config)
+    {
+        isClientsUseV2Format = config.isClientsUseV2Format();
+        this.isGetUserFromOAuth = config.isGetUserFromOauth();
+        if (isGetUserFromOAuth) {
+            this.oauthTokenInfoUrl = requireNonNull(config.getOauthTokenInfoUrl());
+        }
+
+        defaultCatalog = requireNonNullElse(request.getHeader(TRINO_CATALOG_HEADER_NAME), UNSET);
+        defaultSchema = requireNonNullElse(request.getHeader(TRINO_SCHEMA_HEADER_NAME), UNSET);
+        user = extractUser(request, config.getTokenUserField());
+        if (request.getMethod().equals(HttpMethod.POST)) {
+            processRequestBody(request, config);
+        }
+    }
+
+    private void processRequestBody(HttpServletRequest request, ProcessedRequestConfig config)
+    {
+        try (BufferedReader reader = request.getReader()) {
+            if (reader == null) {
+                log.warn("HTTP request returned null reader");
+                body = "";
+                return;
+            }
+
+            Map<String, String> preparedStatements = getPreparedStatements(request);
+            SqlParser parser = new SqlParser();
+            reader.mark(config.getMaxBodySize());
+            body = CharStreams.toString(request.getReader());
+
+            if (isClientsUseV2Format) {
```

Comment:
can this be determined per request basis? looking at the url or any other mechanism?